### PR TITLE
Stripped version

### PR DESCRIPTION
**Pull request by deuszex**
feat-2.0 compatible.
Events:
EventsMode::CES events included (forked original to update to rc3).
EventsMode::Native events included (Casper Native events).
EventsMode::NativeBytes (uses similar bytes as ces).
Compiles on latest rust.

Added 1.5.6 fixture(state snapshot) to test migration from 1.5.6 to feat-2.0.

### CODE REVIEW

File: cep18/src/events.rs [hunk 1]
```diff
@@ -99,14 +115,17 @@ fn ces(event: Event) {
...
 }
 
 pub fn init_events() {
-    let events_mode: EventsMode =
-        EventsMode::try_from(read_from::<u8>(EVENTS_MODE)).unwrap_or_revert();
+    let events_mode: EventsMode = EventsMode::try_from(read_from::<u8>(EVENTS_MODE))
+        .unwrap_or_revert_with(Cep18Error::InvalidEventsMode);
 
-    if events_mode == EventsMode::CES {
+    if [EventsMode::CES].contains(&events_mode)
```
Comment (gRoussac) [inline cep18/src/events.rs]:
```suggestion
    if events_mode == EventsMode::CES

File: cep18/src/main.rs [hunk 2]
```diff
@@ -271,31 +304,52 @@ pub extern "C" fn init() {
...
+        get_optional_named_arg_with_user_errors(ADMIN_LIST, Cep18Error::InvalidAdminList);
     let minter_list: Option<Vec<Key>> =
-        utils::get_optional_named_arg_with_user_errors(MINTER_LIST, Cep18Error::InvalidMinterList);
+        get_optional_named_arg_with_user_errors(MINTER_LIST, Cep18Error::InvalidMinterList);
 
-    init_events();
+    let events_mode: EventsMode = EventsMode::try_from(get_named_arg::<u8>(EVENTS_MODE))
+        .unwrap_or_revert_with(Cep18Error::InvalidEventsMode);
+
+    if [EventsMode::CES].contains(&events_mode) {
```
Comment (gRoussac) [inline cep18/src/main.rs]:
```suggestion
    if events_mode == EventsMode::CES {

File: cep18/src/events.rs (hunk 1 above)
Comment (gRoussac) [inline cep18/src/events.rs]:
```suggestion
    if events_mode == EventsMode::CES

File: cep18/src/main.rs [hunk 3]
```diff
@@ -351,63 +410,231 @@ pub extern "C" fn change_security() {
...
+        .unwrap_or_revert_with(Cep18Error::InvalidEventsMode);
+    let old_events_mode: EventsMode = EventsMode::try_from(read_from::<u8>(EVENTS_MODE))
+        .unwrap_or_revert_with(Cep18Error::InvalidEventsMode);
+    if events_mode == old_events_mode {
+        revert(Cep18Error::UnchangedEventsMode);
+    }
+    let events_mode_u8 = events_mode as u8;
+    put_key(EVENTS_MODE, storage::new_uref(events_mode_u8).into());
+
+    if get_key(casper_event_standard::EVENTS_DICT).is_some() {
```
Comment (gRoussac) [inline cep18/src/main.rs]: should it not be is_none here ?

File: cep18/src/main.rs [hunk 4]
```diff
@@ -425,39 +652,49 @@ pub fn install_contract(name: &str) {
         ENABLE_MINT_BURN.to_string(),
         storage::new_uref(enable_mint_burn).into(),
     );
+
     let entry_points = generate_entry_points();
 
-    let hash_key_name = format!("{HASH_KEY_NAME_PREFIX}{name}");
+    let mut message_topics = BTreeMap::new();
+    if [EventsMode::Native].contains(&events_mode.try_into().unwrap_or_default()) {
```
Comment (gRoussac) [inline cep18/src/main.rs]:
```suggestion 
if events_mode.try_into().unwrap_or_default() == EventsMode::Native {
```

Please check if correct

File: cep18/src/utils.rs [hunk 5]
```diff
@@ -60,11 +65,56 @@ fn call_stack_element_to_address(call_stack_element: CallStackElement) -> Key {
     }
 }
 
+/// Returns the call stack.
+pub fn get_call_stack() -> Vec<CallStackElement> {
```
Comment (gRoussac) [inline cep18/src/utils.rs:69]: what is the difference between runtime::get_call_stack ? thanks

File: tests/src/allowance.rs [hunk 6]
```diff
@@ -53,64 +54,80 @@ fn should_approve_funds_contract_to_contract() {
...
+    let (_, account_user_1_account_hash, _) = get_test_account("ACCOUNT_USER_1");
+    let account_user_1_key = Key::Account(account_user_1_account_hash);
+
     test_approve_for(
         &mut builder,
         &test_context,
-        Key::Account(*DEFAULT_ACCOUNT_ADDR),
-        Key::Account(*DEFAULT_ACCOUNT_ADDR),
-        Key::Account(*ACCOUNT_1_ADDR),
+        default_account_user_key,
```
Comment (gRoussac) [inline tests/src/allowance.rs:63]: seems the same as as *DEFAULT_ACCOUNT_KEY ?

File: tests/src/allowance.rs [hunk 7]
```diff
@@ -53,64 +54,80 @@ fn should_approve_funds_contract_to_contract() {
 fn should_approve_funds_account_to_account() {
     let (mut builder, test_context) = setup();
 
+    let default_account_user_key = Key::Account(*DEFAULT_ACCOUNT_ADDR);
+    let (_, account_user_1_account_hash, _) = get_test_account("ACCOUNT_USER_1");
+    let account_user_1_key = Key::Account(account_user_1_account_hash);
```
Comment (gRoussac) [inline tests/src/allowance.rs]:
```suggestion 
let (account_user_1_key, account_user_1_account_hash, _) = get_test_account("ACCOUNT_USER_1");

File: tests/src/allowance.rs [hunk 8]
```diff
@@ -53,64 +54,80 @@ fn should_approve_funds_contract_to_contract() {
...
-    let (mut builder, TestContext { cep18_token, .. }) = setup();
+    let (
+        mut builder,
+        TestContext {
+            cep18_contract_hash,
+            ..
+        },
+    ) = setup();
+
+    let (_, default_account_user_account_hash, _) = get_test_account("ACCOUNT_USER_0");
```
Comment (gRoussac) [inline tests/src/allowance.rs]:
```suggestion
let (default_account_user_account_key, default_account_user_account_hash, _) = get_test_account("ACCOUNT_USER_0");
```

File: tests/src/allowance.rs [hunk 9]
```diff
@@ -53,64 +54,80 @@ fn should_approve_funds_contract_to_contract() {
...
+    let addressable_cep18_contract_hash = AddressableEntityHash::new(cep18_contract_hash.value());
 
     let allowance_amount_1 = U256::from(ALLOWANCE_AMOUNT_1);
     let transfer_from_amount_1 = allowance_amount_1 + U256::one();
 
-    let sender = *DEFAULT_ACCOUNT_ADDR;
-    let owner = sender;
-    let recipient = *ACCOUNT_1_ADDR;
+    let sender = default_account_user_account_hash;
+    let sender_key = Key::Account(*DEFAULT_ACCOUNT_KEY);
```
Comment (gRoussac) [inline tests/src/allowance.rs:100]:
```suggestion
let sender_key = default_account_user_account_key;

File: tests/src/fixture_gen.rs [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+use once_cell::sync::Lazy;
+
+pub const CEP18_CONTRACT_WASM: &str = "cep18.wasm";
+pub const CEP18_TEST_CONTRACT_WASM: &str = "cep18_test_contract.wasm";
+pub const NAME_KEY: &str = "name";
+pub const SYMBOL_KEY: &str = "symbol";
+pub const CEP18_TOKEN_CONTRACT_KEY: &str = "cep18_contract_hash_CasperTest";
+pub const DECIMALS_KEY: &str = "decimals";
+pub const TOTAL_SUPPLY_KEY: &str = "total_supply";
+pub const BALANCES_KEY: &str = "balances";
```
Comment (gRoussac) [inline tests/src/fixture_gen.rs]: Was it not possible to resuse constants in ? [tests/src/utility/constants.rs](https://github.com/casper-ecosystem/cep18/pull/155/files#diff-4f618a198724cd35804b539b8ada355222704900f7d949eeb083c9146c001eb3)

File: tests/src/mint_and_burn.rs [hunk 11]
```diff
@@ -266,22 +317,31 @@ fn test_should_not_mint_or_burn_with_entrypoint_disabled() {
 
 #[test]
 fn test_security_no_rights() {
+    let (_, account_user_1_account_hash, _) = get_test_account("ACCOUNT_USER_1");
```
Comment (gRoussac) [inline tests/src/mint_and_burn.rs]:
```suggestion
let (account_user_1_key, account_user_1_account_hash, _) = get_test_account("ACCOUNT_USER_1");

File: tests/src/mint_and_burn.rs [hunk 12]
```diff
@@ -266,22 +317,31 @@ fn test_should_not_mint_or_burn_with_entrypoint_disabled() {
...
+        },
+    ) = setup_with_args(runtime_args! {
         ARG_NAME => TOKEN_NAME,
         ARG_SYMBOL => TOKEN_SYMBOL,
         ARG_DECIMALS => TOKEN_DECIMALS,
         ARG_TOTAL_SUPPLY => U256::from(TOKEN_TOTAL_SUPPLY),
         ENABLE_MINT_BURN => true,
     });
-
+    let account_user_1_key = Key::Account(account_user_1_account_hash);
```
Comment (gRoussac) [inline tests/src/mint_and_burn.rs]:
```suggestion

File: tests/src/mint_and_burn.rs [hunk 13]
```diff
@@ -328,23 +388,32 @@ fn test_security_no_rights() {
 
 #[test]
 fn test_security_minter_rights() {
+    let (_, account_user_1_account_hash, _) = get_test_account("ACCOUNT_USER_1");
     let mint_amount = U256::one();
+    let account_user_1_key = Key::Account(account_user_1_account_hash);
```
Comment (gRoussac) [inline tests/src/mint_and_burn.rs]:
```suggestion
    let (account_user_1_key, account_user_1_account_hash, _) = get_test_account("ACCOUNT_USER_1");
    let mint_amount = U256::one();

File: tests/src/allowance.rs (hunk 6 above)
Comment (gRoussac) [reply tests/src/allowance.rs:63]: Can you maybe replace every occurrences in every files  ?
```
Key::Account(*DEFAULT_ACCOUNT_ADDR)
```
by 

```
*DEFAULT_ACCOUNT_KEY
```

File: tests/src/transfer.rs [hunk 14]
```diff
@@ -1,50 +1,60 @@
...
+        mut builder,
+        TestContext {
+            cep18_contract_hash,
+            ..
+        },
+    ) = setup();
+
+    let default_account_user_key = Key::Account(*DEFAULT_ACCOUNT_ADDR);
+    let (_, account_user_1_hash, _) = get_test_account("ACCOUNT_USER_1");
+    let account_user_1_key = Key::Account(account_user_1_hash);
```
Comment (gRoussac) [inline tests/src/transfer.rs]:
```suggestion
    let (account_user_1_key, account_user_1_hash, _) = get_test_account("ACCOUNT_USER_1");

File: tests/src/transfer.rs [hunk 15]
```diff
@@ -56,50 +66,60 @@ fn should_transfer_full_owned_amount() {
...
+        TestContext {
+            cep18_contract_hash,
+            ..
+        },
+    ) = setup();
+
+    let (_, default_account_user_account_hash, _) = get_test_account("ACCOUNT_USER_0");
+    let default_account_user_key = Key::Account(*DEFAULT_ACCOUNT_ADDR);
+    let (_, account_user_1_hash, _) = get_test_account("ACCOUNT_USER_1");
+    let account_user_1_key = Key::Account(account_user_1_hash);
```
Comment (gRoussac) [inline tests/src/transfer.rs]:
```suggestion
    let (default_account_user_key, default_account_user_account_hash, _) = get_test_account("ACCOUNT_USER_0");
    let (account_user_1_key, account_user_1_hash, _) = get_test_account("ACCOUNT_USER_1");

File: tests/src/transfer.rs [hunk 16]
```diff
@@ -114,58 +134,71 @@ fn should_not_transfer_more_than_owned_balance() {
...
+        TestContext {
+            cep18_contract_hash,
+            ..
+        },
+    ) = setup();
+
+    let (_, default_account_user_account_hash, _) = get_test_account("ACCOUNT_USER_0");
+    let default_account_user_key = Key::Account(*DEFAULT_ACCOUNT_ADDR);
+    let (_, account_user_1_account_hash, _) = get_test_account("ACCOUNT_USER_1");
+    let account_user_1_key = Key::Account(account_user_1_account_hash);
```
Comment (gRoussac) [inline tests/src/transfer.rs]:
```suggestion
    let (default_account_user_key, default_account_user_account_hash, _) = get_test_account("ACCOUNT_USER_0");
    let (account_user_1_key, account_user_1_account_hash, _) = get_test_account("ACCOUNT_USER_1");

File: tests/src/transfer.rs [hunk 17]
```diff
@@ -206,40 +239,46 @@ fn should_transfer_from_account_by_contract() {
...
-            cep18_token,
+            cep18_contract_hash,
             cep18_test_contract_package,
             ..
         },
     ) = setup();
 
+    let default_account_user_key = Key::Account(*DEFAULT_ACCOUNT_ADDR);
+    let (_, account_user_1_account_hash, _) = get_test_account("ACCOUNT_USER_1");
+    let account_user_1_key = Key::Account(account_user_1_account_hash);
```
Comment (gRoussac) [inline tests/src/transfer.rs]:
```suggestion
    let default_account_user_key = Key::Account(*DEFAULT_ACCOUNT_ADDR);
    let (account_user_1_key, account_user_1_account_hash, _) = get_test_account("ACCOUNT_USER_1");

File: tests/src/transfer.rs [hunk 18]
```diff
@@ -313,18 +361,30 @@ fn should_not_be_able_to_own_transfer() {
...
+            ..
+        },
+    ) = setup();
 
-    let owner = Key::Account(*DEFAULT_ACCOUNT_ADDR);
-    let spender = Key::Account(*DEFAULT_ACCOUNT_ADDR);
-    let sender = Key::Account(*DEFAULT_ACCOUNT_ADDR);
-    let recipient = Key::Account(*DEFAULT_ACCOUNT_ADDR);
+    let (_, default_account_account_hash, _) = get_test_account("ACCOUNT_USER_0");
+    let default_account_user_key = Key::Account(*DEFAULT_ACCOUNT_ADDR);
```
Comment (gRoussac) [inline tests/src/transfer.rs]:
```suggestion
    let (default_account_user_key, default_account_account_hash, _) = get_test_account("ACCOUNT_USER_0");

File: tests/src/transfer.rs [hunk 19]
```diff
@@ -367,61 +429,90 @@ fn should_not_be_able_to_own_transfer_from() {
...
+            cep18_contract_hash,
+            ..
+        },
+    ) = setup();
 
-    let sender = Key::Account(*DEFAULT_ACCOUNT_ADDR);
-    let recipient = Key::Account(*ACCOUNT_1_ADDR);
+    let default_account_user_key = Key::Account(*DEFAULT_ACCOUNT_ADDR);
+    let (_, account_user_1_account_hash, _) = get_test_account("ACCOUNT_USER_1");
+    let account_user_1_key = Key::Account(account_user_1_account_hash);
```
Comment (gRoussac) [inline tests/src/transfer.rs]:
```suggestion
    let default_account_user_key = Key::Account(*DEFAULT_ACCOUNT_ADDR);
    let (account_user_1_key, account_user_1_account_hash, _) = get_test_account("ACCOUNT_USER_1");

File: tests/src/transfer.rs [hunk 20]
```diff
@@ -367,61 +429,90 @@ fn should_not_be_able_to_own_transfer_from() {
...
+    ) = setup();
 
-    let owner = Key::Account(*DEFAULT_ACCOUNT_ADDR);
-    let spender = Key::Account(*ACCOUNT_1_ADDR);
-    let recipient = Key::Account(*ACCOUNT_2_ADDR);
+    let default_account_user_key = Key::Account(*DEFAULT_ACCOUNT_ADDR);
+    let (_, account_user_1_account_hash, _) = get_test_account("ACCOUNT_USER_1");
+    let account_user_1_key = Key::Account(account_user_1_account_hash);
+    let (_, account_user_2_account_hash, _) = get_test_account("ACCOUNT_USER_2");
+    let account_user_2_key = Key::Account(account_user_2_account_hash);
```
Comment (gRoussac) [inline tests/src/transfer.rs]:
```suggestion
    let default_account_user_key = *DEFAULT_ACCOUNT_KEY;
    let (account_user_1_key, account_user_1_account_hash, _) = get_test_account("ACCOUNT_USER_1");
    let (account_user_2_key, account_user_2_account_hash, _) = get_test_account("ACCOUNT_USER_2");

File: tests/src/transfer.rs [hunk 21]
```diff
@@ -474,11 +568,15 @@ fn should_transfer_contract_to_account() {
         ..
     } = test_context;
 
-    let sender1 = Key::Account(*DEFAULT_ACCOUNT_ADDR);
+    let default_account_user_key = Key::Account(*DEFAULT_ACCOUNT_ADDR);
+    let (_, account_user_1_account_hash, _) = get_test_account("ACCOUNT_USER_1");
+    let account_user_1_key = Key::Account(account_user_1_account_hash);
```
Comment (gRoussac) [inline tests/src/transfer.rs]:
```suggestion
    let default_account_user_key = *DEFAULT_ACCOUNT_KEY;
    let (account_user_1_key, account_user_1_account_hash, _) = get_test_account("ACCOUNT_USER_1");

File: tests/src/transfer.rs [hunk 22]
```diff
@@ -494,9 +592,13 @@ fn should_transfer_contract_to_account() {
 fn should_transfer_account_to_contract() {
     let (mut builder, test_context) = setup();
 
-    let sender1 = Key::Account(*DEFAULT_ACCOUNT_ADDR);
-    let recipient1 = Key::Account(*ACCOUNT_1_ADDR);
-    let sender2 = Key::Account(*ACCOUNT_1_ADDR);
+    let default_account_user_key = Key::Account(*DEFAULT_ACCOUNT_ADDR);
+    let (_, account_user_1_account_hash, _) = get_test_account("ACCOUNT_USER_1");
+    let account_user_1_key = Key::Account(account_user_1_account_hash);
```
Comment (gRoussac) [inline tests/src/transfer.rs]:
```suggestion
    let default_account_user_key = *DEFAULT_ACCOUNT_KEY;
    let (account_user_1_key, account_user_1_account_hash, _) = get_test_account("ACCOUNT_USER_1");

File: tests/src/transfer.rs [hunk 23]
```diff
@@ -512,10 +614,17 @@ fn should_transfer_account_to_contract() {
...
-    let sender1 = Key::Account(*DEFAULT_ACCOUNT_ADDR);
-    let recipient1 = Key::Account(*ACCOUNT_1_ADDR);
-    let sender2 = Key::Account(*ACCOUNT_1_ADDR);
-    let recipient2 = Key::Account(*ACCOUNT_2_ADDR);
+
+    let default_account_user_key = Key::Account(*DEFAULT_ACCOUNT_ADDR);
+    let (_, account_user_1_account_hash, _) = get_test_account("ACCOUNT_USER_1");
+    let account_user_1_key = Key::Account(account_user_1_account_hash);
+    let (_, account_user_2_account_hash, _) = get_test_account("ACCOUNT_USER_2");
+    let account_user_2_key = Key::Account(account_user_2_account_hash);
```
Comment (gRoussac) [inline tests/src/transfer.rs]:
```suggestion
    let default_account_user_key = *DEFAULT_ACCOUNT_KEY;
    let (account_user_1_key, account_user_1_account_hash, _) = get_test_account("ACCOUNT_USER_1");
    let (account_user_2_key, account_user_2_account_hash, _) = get_test_account("ACCOUNT_USER_2");

**Review by gRoussac — APPROVED** [review]
I left couple of little nips, please check if still compiles after suggestions if applied.

File: cep18/src/utils.rs (hunk 5 above)
Comment (deuszex) [reply cep18/src/utils.rs:69]: The current system implementation of `get_call_stack` uses Key::Package and such, which we unfortunately have to circumvent here, by relying on a long deprecated FFI call.

File: cep18/src/main.rs (hunk 4 above)
Comment (deuszex) [reply cep18/src/main.rs]: Thanks for the catch it actually wasn't needed an extra modality there

File: tests/src/allowance.rs (hunk 6 above)
Comment (deuszex) [reply tests/src/allowance.rs:63]: Nope, currently [DEFAULT_ACCOUNT_KEY is the exact same value as DEFAULT_ACCOUNT_ADDR](https://github.com/casper-network/casper-node/blob/dev/execution_engine_testing/test_support/src/lib.rs#L97), so it's not a key but an AccountHash

**Pull request merged**